// beachball/lib/__e2e__/validation.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const repository_1 = require("../fixtures/repository");
const isChangeFileNeeded_1 = require("../validation/isChangeFileNeeded");
const writeChangeFiles_1 = require("../changefile/writeChangeFiles");
const areChangeFilesDeleted_1 = require("../validation/areChangeFilesDeleted");
const paths_1 = require("../paths");
const fs_extra_1 = __importDefault(require("fs-extra"));
describe('validation', () => {
    let repositoryFactory;
    beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
        repositoryFactory = new repository_1.RepositoryFactory();
        yield repositoryFactory.create();
    }));
    afterAll(() => __awaiter(void 0, void 0, void 0, function* () {
        yield repositoryFactory.cleanUp();
    }));
    describe('isChangeFileNeeded', () => {
        let repository;
        beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
            repository = yield repositoryFactory.cloneRepository();
        }));
        it('is false when no changes have been made', () => __awaiter(void 0, void 0, void 0, function* () {
            const result = isChangeFileNeeded_1.isChangeFileNeeded({
                branch: 'origin/master',
                path: repository.rootPath,
                fetch: false,
            });
            expect(result).toBeFalsy();
        }));
        it('is true when changes exist in a new branch', () => __awaiter(void 0, void 0, void 0, function* () {
            yield repository.branch('feature-0');
            yield repository.commitChange('myFilename');
            const result = isChangeFileNeeded_1.isChangeFileNeeded({
                branch: 'origin/master',
                path: repository.rootPath,
                fetch: false,
            });
            expect(result).toBeTruthy();
        }));
        it('is false when changes are CHANGELOG files', () => __awaiter(void 0, void 0, void 0, function* () {
            yield repository.branch('feature-0');
            yield repository.commitChange('CHANGELOG.md');
            const result = isChangeFileNeeded_1.isChangeFileNeeded({
                branch: 'origin/master',
                path: repository.rootPath,
                fetch: false,
            });
            expect(result).toBeFalsy();
        }));
        it('throws if the remote is invalid', () => __awaiter(void 0, void 0, void 0, function* () {
            yield repository.setRemoteUrl('origin', 'file:///__nonexistent');
            yield repository.branch('feature-0');
            yield repository.commitChange('CHANGELOG.md');
            expect(() => {
                isChangeFileNeeded_1.isChangeFileNeeded({
                    branch: 'origin/master',
                    path: repository.rootPath,
                    fetch: true,
                });
            }).toThrow();
        }));
    });
    describe('areChangeFilesDeleted', () => {
        let repository;
        beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
            repository = yield repositoryFactory.cloneRepository();
            writeChangeFiles_1.writeChangeFiles({
                'pkg-1': {
                    type: 'minor',
                    comment: 'test',
                    email: 'test@test.com',
                    packageName: 'pkg-1',
                    dependentChangeType: 'patch',
                },
            }, repository.rootPath);
            yield repository.push('origin', 'master');
        }));
        it('is false when no change files are deleted', () => __awaiter(void 0, void 0, void 0, function* () {
            yield repository.branch('feature-0');
            const result = areChangeFilesDeleted_1.areChangeFilesDeleted({
                branch: 'origin/master',
                path: repository.rootPath,
            });
            expect(result).toBeFalsy();
        }));
        it('is true when change files are deleted', () => __awaiter(void 0, void 0, void 0, function* () {
            var _a;
            yield repository.branch('feature-0');
            const changeDirPath = (_a = paths_1.getChangePath(repository.rootPath), (_a !== null && _a !== void 0 ? _a : fail('No change folder found')));
            fs_extra_1.default.removeSync(changeDirPath);
            yield repository.commitAll();
            const result = areChangeFilesDeleted_1.areChangeFilesDeleted({
                branch: 'origin/master',
                path: repository.rootPath,
            });
            expect(result).toBeTruthy();
        }));
    });
});
