// beachball/lib/options/getCliOptions.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const yargs_parser_1 = __importDefault(require("yargs-parser"));
const paths_1 = require("../paths");
const workspace_tools_1 = require("workspace-tools");
let cachedCliOptions;
function getCliOptions(argv) {
    // Special case caching to process.argv which should be immutable
    if (argv === process.argv) {
        if (!cachedCliOptions) {
            cachedCliOptions = getCliOptionsUncached(process.argv);
        }
        return cachedCliOptions;
    }
    else {
        return getCliOptionsUncached(argv);
    }
}
exports.getCliOptions = getCliOptions;
function getCliOptionsUncached(argv) {
    // Be careful not to mutate the input argv
    const trimmedArgv = [...argv].splice(2);
    const args = yargs_parser_1.default(trimmedArgv, {
        string: ['branch', 'tag', 'message', 'package', 'since', 'dependent-change-type', 'config'],
        array: ['scope', 'disallowed-change-types'],
        boolean: ['git-tags', 'keep-change-files', 'force', 'disallow-deleted-change-files', 'no-commit'],
        alias: {
            branch: ['b'],
            config: ['c'],
            tag: ['t'],
            registry: ['r'],
            message: ['m'],
            token: ['n'],
            help: ['h', '?'],
            yes: ['y'],
            package: ['p'],
            version: ['v'],
        },
    });
    const { _ } = args, restArgs = __rest(args, ["_"]);
    const cwd = paths_1.findProjectRoot(process.cwd()) || process.cwd();
    const cliOptions = Object.assign(Object.assign(Object.assign({}, (_.length > 0 && { command: _[0] })), restArgs), { path: cwd, fromRef: args.since, keepChangeFiles: args['keep-change-files'], disallowDeletedChangeFiles: args['disallow-deleted-change-files'], forceVersions: args.force, configPath: args.config });
    const disallowedChangeTypesArgs = args['disallowed-change-types'];
    if (disallowedChangeTypesArgs) {
        cliOptions.disallowedChangeTypes = disallowedChangeTypesArgs;
    }
    if (args.branch) {
        cliOptions.branch = args.branch.indexOf('/') > -1 ? args.branch : workspace_tools_1.getDefaultRemoteBranch(args.branch, cwd);
    }
    return cliOptions;
}
